The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3052808759/jobs/4922627797) [(plain)](https://github.com/rust-lang/rust/commit/f0922dfebbd546bfcc2484994bbee713bcb63064/checks/8350231255/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rand v0.7.3
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking std v0.0.0 (/checkout/library/std)
    Checking core v0.0.0 (/checkout/library/core)
error[E0412]: cannot find type `RawWakerVTable` in this scope
   |
   |
18 |     const VOID_TABLE: RawWakerVTable = RawWakerVTable::new(|_| VOID_WAKER, |_| {}, |_| {}, |_| {});
   |
help: consider importing one of these items
   |
   |
1  | use core::task::RawWakerVTable;
   |
1  | use std::task::RawWakerVTable;


error[E0433]: failed to resolve: use of undeclared type `RawWakerVTable`
   |
   |
18 |     const VOID_TABLE: RawWakerVTable = RawWakerVTable::new(|_| VOID_WAKER, |_| {}, |_| {}, |_| {});
   |
help: consider importing one of these items
   |
   |
1  | use core::task::RawWakerVTable;
   |
1  | use std::task::RawWakerVTable;


error[E0412]: cannot find type `RawWaker` in this scope
   |
   |
20 |     const VOID_WAKER: RawWaker = RawWaker::new(&(), &VOID_TABLE);
   |
help: consider importing one of these items
   |
   |
1  | use core::task::RawWaker;
   |
1  | use std::task::RawWaker;


error[E0433]: failed to resolve: use of undeclared type `RawWaker`
   |
   |
20 |     const VOID_WAKER: RawWaker = RawWaker::new(&(), &VOID_TABLE);
   |
help: consider importing one of these items
   |
   |
1  | use core::task::RawWaker;
   |
1  | use std::task::RawWaker;

error[E0412]: cannot find type `Waker` in this scope
  --> library/core/tests/task.rs:22:18
   |
   |
22 |     const WAKER: Waker = unsafe { Waker::from_raw(VOID_WAKER) };
   |
help: consider importing one of these items
   |
1  | use core::task::Waker;
1  | use core::task::Waker;
   |
1  | use std::task::Waker;
   |

error[E0433]: failed to resolve: use of undeclared type `Waker`
  --> library/core/tests/task.rs:22:35
   |
22 |     const WAKER: Waker = unsafe { Waker::from_raw(VOID_WAKER) };
   |
help: consider importing one of these items
   |
1  | use core::task::Waker;
1  | use core::task::Waker;
   |
1  | use std::task::Waker;
   |

error[E0412]: cannot find type `Context` in this scope
  --> library/core/tests/task.rs:24:20
   |
24 |     const CONTEXT: Context<'static> = Context::from_waker(&WAKER);
   |
help: consider importing one of these items
   |
1  | use core::task::Context;
1  | use core::task::Context;
   |
1  | use std::task::Context;
   |

error[E0433]: failed to resolve: use of undeclared type `Context`
  --> library/core/tests/task.rs:24:39
   |
24 |     const CONTEXT: Context<'static> = Context::from_waker(&WAKER);
   |
help: consider importing one of these items
   |
1  | use core::task::Context;
1  | use core::task::Context;
   |
1  | use std::task::Context;
   |

error[E0412]: cannot find type `Waker` in this scope
  --> library/core/tests/task.rs:26:31
   |
26 |     const WAKER_REF: &'static Waker = CONTEXT.waker();
   |
help: consider importing one of these items
   |
1  | use core::task::Waker;
```

</details>

--- a/library/alloc/src/lib.rs
+++ b/library/alloc/src/lib.rs
@@ -109,6 +109,7 @@
 #![feature(core_intrinsics)]
 #![feature(const_eval_select)]
 #![feature(const_pin)]
+#![feature(const_waker)]
 #![feature(cstr_from_bytes_until_nul)]
 #![feature(dispatch_from_dyn)]
 #![cfg_attr(not(bootstrap), feature(error_generic_member_access))]

--- a/library/core/src/lib.rs
+++ b/library/core/src/lib.rs
@@ -145,6 +145,7 @@
 #![feature(const_default_impls)]
 #![feature(const_unicode_case_lookup)]
 #![feature(const_unsafecell_get_mut)]
+#![feature(const_waker)]
 #![feature(core_panic)]
 #![feature(duration_consts_float)]
 #![feature(maybe_uninit_uninit_array)]

--- a/library/core/src/task/wake.rs
+++ b/library/core/src/task/wake.rs
@@ -186,7 +186,7 @@ pub struct Context<'a> {
 impl<'a> Context<'a> {
     /// Create a new `Context` from a [`&Waker`](Waker).
     #[stable(feature = "futures_api", since = "1.36.0")]
-    #[rustc_const_unstable(feature = "const_waker", issue = "none")]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
     #[must_use]
     #[inline]
     pub const fn from_waker(waker: &'a Waker) -> Self {
@@ -195,7 +195,7 @@ impl<'a> Context<'a> {
 
     /// Returns a reference to the [`Waker`] for the current task.
     #[stable(feature = "futures_api", since = "1.36.0")]
-    #[rustc_const_unstable(feature = "const_waker", issue = "none")]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
     #[must_use]
     #[inline]
     pub const fn waker(&self) -> &'a Waker {
@@ -313,7 +313,7 @@ impl Waker {
     #[inline]
     #[must_use]
     #[stable(feature = "futures_api", since = "1.36.0")]
-    #[rustc_const_unstable(feature = "const_waker", issue = "none")]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
     pub const unsafe fn from_raw(waker: RawWaker) -> Waker {
         Waker { waker }
     }

--- a/library/core/src/task/wake.rs
+++ b/library/core/src/task/wake.rs
@@ -186,17 +186,19 @@ pub struct Context<'a> {
 impl<'a> Context<'a> {
     /// Create a new `Context` from a [`&Waker`](Waker).
     #[stable(feature = "futures_api", since = "1.36.0")]
+    #[rustc_const_unstable(feature = "const_waker", issue = "none")]
     #[must_use]
     #[inline]
-    pub fn from_waker(waker: &'a Waker) -> Self {
+    pub const fn from_waker(waker: &'a Waker) -> Self {
         Context { waker, _marker: PhantomData }
     }
 
     /// Returns a reference to the [`Waker`] for the current task.
     #[stable(feature = "futures_api", since = "1.36.0")]
+    #[rustc_const_unstable(feature = "const_waker", issue = "none")]
     #[must_use]
     #[inline]
-    pub fn waker(&self) -> &'a Waker {
+    pub const fn waker(&self) -> &'a Waker {
         &self.waker
     }
 }
@@ -311,7 +313,8 @@ impl Waker {
     #[inline]
     #[must_use]
     #[stable(feature = "futures_api", since = "1.36.0")]
-    pub unsafe fn from_raw(waker: RawWaker) -> Waker {
+    #[rustc_const_unstable(feature = "const_waker", issue = "none")]
+    pub const unsafe fn from_raw(waker: RawWaker) -> Waker {
         Waker { waker }
     }
 

--- a/library/core/tests/lib.rs
+++ b/library/core/tests/lib.rs
@@ -74,6 +74,7 @@
 #![feature(iterator_try_reduce)]
 #![feature(const_mut_refs)]
 #![feature(const_pin)]
+#![feature(const_waker)]
 #![feature(never_type)]
 #![feature(unwrap_infallible)]
 #![feature(pointer_byte_offsets)]

--- a/library/core/tests/task.rs
+++ b/library/core/tests/task.rs
@@ -1,4 +1,4 @@
-use core::task::Poll;
+use core::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
 
 #[test]
 fn poll_const() {
@@ -11,4 +11,19 @@ fn poll_const() {
 
     const IS_PENDING: bool = POLL.is_pending();
     assert!(IS_PENDING);
+}
+
+#[test]
+fn waker_const() {
+    const VOID_TABLE: RawWakerVTable = RawWakerVTable::new(|_| VOID_WAKER, |_| {}, |_| {}, |_| {});
+
+    const VOID_WAKER: RawWaker = RawWaker::new(&(), &VOID_TABLE);
+
+    static WAKER: Waker = unsafe { Waker::from_raw(VOID_WAKER) };
+
+    static CONTEXT: Context<'static> = Context::from_waker(&WAKER);
+
+    static WAKER_REF: &'static Waker = CONTEXT.waker();
+
+    WAKER_REF.wake_by_ref();
 }

--- a/library/std/src/lib.rs
+++ b/library/std/src/lib.rs
@@ -315,6 +315,7 @@
 #![feature(strict_provenance)]
 #![feature(maybe_uninit_uninit_array)]
 #![feature(const_maybe_uninit_uninit_array)]
+#![feature(const_waker)]
 //
 // Library features (alloc):
 #![feature(alloc_layout_extra)]

